File: tinkerpop3/src/test/java/overflowdb/testdomains/gratefuldead/GratefulDead.java
package overflowdb.testdomains.gratefuldead;

import org.apache.tinkerpop.gremlin.structure.io.IoCore;
import overflowdb.Config;
import overflowdb.Graph;
import overflowdb.tinkerpop.OdbGraphTp3;

import java.io.IOException;
import java.util.Arrays;

public class GratefulDead {
  public static Graph open() {
    return open(Config.withoutOverflow());
  }

  public static Graph open(Config config) {
    return Graph.open(
        config,
        Arrays.asList(Song.factory, Artist.factory),
        Arrays.asList(FollowedBy.factory, SungBy.factory, WrittenBy.factory)
    );
  }

  public static OdbGraphTp3 openAndLoadSampleData() throws IOException {
    OdbGraphTp3 graph = OdbGraphTp3.wrap(open());
    loadData(graph);
    return graph;
  }

  public static OdbGraphTp3 openAndLoadSampleData(String path) throws IOException {
    OdbGraphTp3 graph =
      OdbGraphTp3.wrap(
        open(Config.withDefaults().withStorageLocation(path))
      );
    loadData(graph);
    return graph;
  }


  public static void loadData(OdbGraphTp3 graph) throws IOException {
    graph.io(IoCore.graphml()).readGraph("../src/test/resources/grateful-dead.xml");
  }


}

File: traversal/src/test/scala/overflowdb/traversal/testdomains/gratefuldead/GratefulDead.scala
package overflowdb.traversal.testdomains.gratefuldead

import java.util

import overflowdb.tinkerpop.OdbGraphTp3
import overflowdb.traversal.{Traversal, TraversalSource}
import overflowdb.{Config, Graph}

/* visual schema: https://raw.githubusercontent.com/apache/tinkerpop/347eb9a8231c48aa22eced9b07dd6241305898c6/docs/static/images/grateful-dead-schema.png */
object GratefulDead {
  def newGraph: Graph = newGraph(Config.withoutOverflow)

  def newGraph(config: Config): Graph =
    Graph.open(config,
      util.Arrays.asList(Song.factory, Artist.factory),
      util.Arrays.asList(FollowedBy.factory, SungBy.factory, WrittenBy.factory))

  def newGraphWithData: Graph = {
    val graph = newGraph
    loadData(graph)
    graph
  }

  def loadData(graph: Graph): Unit =
    OdbGraphTp3.wrap(graph).traversal().io("src/test/resources/grateful-dead.xml").read().iterate()

  def traversal(graph: Graph) = new GratefulDeadTraversalSource(graph)
}

class GratefulDeadTraversalSource(graph: Graph) extends TraversalSource(graph) {
  def artists: Traversal[Artist] = label(Artist.Label).cast[Artist]
  def songs: Traversal[Song] = label(Song.Label).cast[Song]
}
